core/src/main/scala/com/avast/clients/rabbitmq/StreamingRabbitMQConsumer.scala

```diff
@@ -0,0 +1,272 @@
+package com.avast.clients.rabbitmq
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+
+import cats.effect.concurrent.{Deferred, Semaphore}
+import cats.effect.{Blocker, CancelToken, ConcurrentEffect, ContextShift, Effect, ExitCase, IO, Resource, Sync}
+import cats.syntax.all._
+import com.avast.bytes.Bytes
+import com.avast.clients.rabbitmq.DefaultRabbitMQConsumer.RepublishOriginalRoutingKeyHeaderName
+import com.avast.clients.rabbitmq.DefaultStreamingRabbitMQConsumer.DeliveryQueue
+import com.avast.clients.rabbitmq.api.{Delivery, DeliveryResult}
+import com.avast.metrics.scalaapi.Monitor
+import com.rabbitmq.client._
+import com.typesafe.scalalogging.StrictLogging
+import fs2.Stream
+import fs2.concurrent.Queue
+
+import scala.language.higherKinds
+import scala.util.control.NonFatal
+
+trait StreamingRabbitMQConsumer[F[_], A] {
+  def deliveryStream: fs2.Stream[F, StreamedDelivery[F, A]]
+}
+
+trait StreamedDelivery[+F[_], +A] {
+  def delivery: Delivery[A]
+
+  def handle(result: DeliveryResult): F[StreamedResult]
+}
+
+sealed trait StreamedResult
+object StreamedResult extends StreamedResult
+
+class DefaultStreamingRabbitMQConsumer[F[_]: ConcurrentEffect, A: DeliveryConverter] private (
+    name: String,
+    queueName: String,
+    initialConsumerTag: String,
+    connectionInfo: RabbitMQConnectionInfo,
+    consumerListener: ConsumerListener,
+    monitor: Monitor,
+    blocker: Blocker)(createQueue: F[DeliveryQueue[F, Bytes]], newChannel: F[ServerChannel])(implicit cs: ContextShift[F])
+    extends StreamingRabbitMQConsumer[F, A]
+    with StrictLogging {
+
+  private val streamFailureMeter = monitor.meter("streamFailures")
+
+  private val F: Sync[F] = Sync[F]
+
+  val deliveryStream: fs2.Stream[F, StreamedDelivery[F, A]] = {
+    Stream
+      .eval(recoverIfNeeded)
+      .flatMap { queue =>
+        queue.dequeue
+          .map {
+            case (del, deff) =>
+              new StreamedDelivery[F, A] {
+                override val delivery: Delivery[A] = DefaultRabbitMQClientFactory.convertDelivery(del)
+                override def handle(result: DeliveryResult): F[StreamedResult] = deff.complete(result).map(_ => StreamedResult)
+              }
+          }
+      }
+      .onFinalizeCase(handleStreamFinalize)
+  }
+
+  private val consumer = new AtomicReference[StreamingConsumer]()
+
+  private val isOk = new AtomicBoolean(false)
```

Comment:
There is `cats.effect.Ref` which would probably work better.

Reply:
I didn't know about it however I'm not very keen to refactor it now, as it may bring some more problems which I've already solved :-) So unless you are **strongly** against using `AtomicBoolean`, I'd like to stay with it.

Reply:
I think it is strongly related to the piece of code I want to discuss personally so let's postpone this - but yes, I think you should use `Ref`.